In [ ]:
# IMPORTANT: SOME KAGGLE DATA SOURCES ARE PRIVATE
# RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES.
import kagglehub
kagglehub.login()


Kaggle credentials set.
Kaggle credentials successfully validated.


In [ ]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.

wcukierski_enron_email_dataset_path = kagglehub.dataset_download('wcukierski/enron-email-dataset')
ganiyuolalekan_spam_assassin_email_classification_dataset_path = kagglehub.dataset_download('ganiyuolalekan/spam-assassin-email-classification-dataset')
hinepo_alpaca_dataset_path = kagglehub.dataset_download('hinepo/alpaca-dataset')
# divyamgupta04_enron_struct_data_path = kagglehub.dataset_download('divyamgupta04/enron-struct-data')
divyamgupta04_train_enron_path = kagglehub.dataset_download('divyamgupta04/train-enron')

print('Data source import complete.')


Using Colab cache for faster access to the 'enron-email-dataset' dataset.
Using Colab cache for faster access to the 'spam-assassin-email-classification-dataset' dataset.
Using Colab cache for faster access to the 'alpaca-dataset' dataset.
Using Colab cache for faster access to the 'train-enron' dataset.
Data source import complete.


In [ ]:
import numpy as np
import pandas as pd

import random
import json
import re


In [ ]:
# Install dependencies as needed:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load
file_path = "emails.csv"

# Load the latest version
enron = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "wcukierski/enron-email-dataset",
  file_path,
)

print("First 5 records:", enron.head())

/tmp/ipykernel_55/1993295312.py:10: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  enron = kagglehub.load_dataset(


First 5 records:                        file                                            message
0     allen-p/_sent_mail/1.  Message-ID: <18782981.1075855378110.JavaMail.e...
1    allen-p/_sent_mail/10.  Message-ID: <15464986.1075855378456.JavaMail.e...
2   allen-p/_sent_mail/100.  Message-ID: <24216240.1075855687451.JavaMail.e...
3  allen-p/_sent_mail/1000.  Message-ID: <13505866.1075863688222.JavaMail.e...
4  allen-p/_sent_mail/1001.  Message-ID: <30922949.1075863688243.JavaMail.e...


In [ ]:
# Install dependencies as needed:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load
file_path = "spam_assassin.csv"

# Load the latest version
assassin = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "ganiyuolalekan/spam-assassin-email-classification-dataset",
  file_path,
)

print("First 5 records:", assassin.head())

/tmp/ipykernel_55/273448957.py:10: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  assassin = kagglehub.load_dataset(


First 5 records:                                                 text  target
0  From ilug-admin@linux.ie Mon Jul 29 11:28:02 2...       0
1  From gort44@excite.com Mon Jun 24 17:54:21 200...       1
2  From fork-admin@xent.com Mon Jul 29 11:39:57 2...       1
3  From dcm123@btamail.net.cn Mon Jun 24 17:49:23...       1
4  From ilug-admin@linux.ie Mon Aug 19 11:02:47 2...       0


In [ ]:
# Install dependencies as needed:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load
file_path = "alpaca-dataset.parquet"

# Load the latest version
alpaca = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "hinepo/alpaca-dataset",
  file_path,
)

print("First 5 records:", alpaca.head())

/tmp/ipykernel_55/2087330855.py:10: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  alpaca = kagglehub.load_dataset(


First 5 records:                                          instruction input  \
0               Give three tips for staying healthy.         
1                 What are the three primary colors?         
2                 Describe the structure of an atom.         
3                   How can we reduce air pollution?         
4  Describe a time when you had to make a difficu...         

                                              output  \
0  1.Eat a balanced diet and make sure to include...   
1  The three primary colors are red, blue, and ye...   
2  An atom is made up of a nucleus, which contain...   
3  There are a number of ways to reduce air pollu...   
4  I had to make a difficult decision when I was ...   

                                                text  
0  Below is an instruction that describes a task....  
1  Below is an instruction that describes a task....  
2  Below is an instruction that describes a task....  
3  Below is an instruction that describes a task.... 

In [ ]:
import random
enron.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517401 entries, 0 to 517400
Data columns (total 2 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   file     517401 non-null  object
 1   message  517401 non-null  object
dtypes: object(2)
memory usage: 7.9+ MB


In [ ]:
enron_np = np.array(enron)

In [ ]:
n = random.randint(0,517401)
enron_np[n][1]


'Message-ID: <22189079.1075858183887.JavaMail.evans@thyme>\nDate: Thu, 3 May 2001 07:11:00 -0700 (PDT)\nFrom: george.ellis@americas.bnpparibas.com\nTo: george.ellis@americas.bnpparibas.com\nSubject: Coal\nMime-Version: 1.0\nContent-Type: text/plain; charset=us-ascii\nContent-Transfer-Encoding: 7bit\nX-From: george.ellis@americas.bnpparibas.com\nX-To: george.ellis@americas.bnpparibas.com\nX-cc: \nX-bcc: \nX-Folder: \\Michael_Maggi_Jun2001\\Notes Folders\\All documents\nX-Origin: Maggi-M\nX-FileName: mmaggi.nsf\n\n\n\n    By Timothy Gardner\n    NEW YORK, May 3 (Reuters) - Adding stress to an already\nfrazzled power market, a nationwide coal shortage could spawn\nelectricity brownouts in northeastern states and in the West\nthis summer, experts said on Thursday.\n    While coal provides the country with 51 percent of its\npower, years of industry underinvestment and record prices for\ncompeting fuels natural gas and fuel oil this year have pushed\nutilities\' coal stocks to their lowest 

In [ ]:
print(enron_np[n][1])


Message-ID: <23279079.1075854739128.JavaMail.evans@thyme>
Date: Mon, 8 Jan 2001 04:23:00 -0800 (PST)
From: eric.bass@enron.com
To: jbass1@houston.rr.com
Subject: Re: My Resume
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: Eric Bass
X-To: "Jason Bass" <jbass1@houston.rr.com> @ ENRON
X-cc: 
X-bcc: 
X-Folder: \Eric_Bass_Jun2001\Notes Folders\All documents
X-Origin: Bass-E
X-FileName: ebass.nsf

we would like to meet @ 8 at some Pappasitos (I don't know where you live 
exactly) between you and I.




"Jason Bass" <jbass1@houston.rr.com> on 01/08/2001 11:10:29 AM
To: <Eric.Bass@enron.com>
cc:  
Subject: Re: My Resume


We will be there.  Name the time and which Pappasitos

----- Original Message -----
From: <Eric.Bass@enron.com>
To: <jbass1@houston.rr.com>
Sent: Monday, January 08, 2001 10:46 AM
Subject: Re: My Resume




(See attached file: JBass Resume.doc)
I made a few changes.  Generally when creating a resume, shorter is better
(pr

In [ ]:
import re

def parse_enron_email_tmp(raw_email):
    # Extract Subject
    subject_match = re.search(r"Subject:(.*)", raw_email)
    subject = subject_match.group(1).strip() if subject_match else ""
    # Extract Body (after first blank line)
    parts = raw_email.split("\n\n",1)

    # print(parts)
    body = parts[1] if len(parts) > 1 else ""
    # Remove extra whitespace
    body = re.sub(r"\s+", " ", body).strip()

    # -----------------------------------
    body = "\n".join(
        line for line in body.split("\n")
        if not line.strip().startswith(">")
    )
    # Remove forwarded/reply chains
    body = re.split(r"-----Original Message-----", body)[0]
    body = re.split(r"----------- Forwarded by", body)[0]

    # Remove signatures (basic)
    body = re.split(r"(Regards|Thanks|Sincerely)", body)[0]
    # Final cleanup
    body = re.sub(r"\s+", " ", body).strip()
    part = raw_email.split("\n")
    # print(part)
    metadata=''
    for i in range (1,len(part)):
        if(len(part[i]) > 1 and part[i]!=''):
            metadata = metadata + part[i]

    return subject, body, metadata

In [ ]:
subject, body, metadata = parse_enron_email_tmp(enron_np[100][1])
print(metadata)

Date: Wed, 16 Aug 2000 05:35:00 -0700 (PDT)From: phillip.allen@enron.comTo: cooper.richey@enron.comSubject: Re:Mime-Version: 1.0Content-Type: text/plain; charset=us-asciiContent-Transfer-Encoding: 7bitX-From: Phillip K AllenX-To: Cooper RicheyX-cc: X-bcc: X-Folder: \Phillip_Allen_Dec2000\Notes Folders\'sent mailX-Origin: Allen-PX-FileName: pallen.nsfI tried the new address but I don't have access.  also, what do I need to enter under domain?


In [ ]:
import re

def parse_enron_email(raw_email):
    # print(raw_email)
    # print(";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;")
    # Extract Subject
    subject_match = re.search(r"Subject:(.*)", raw_email)
    # print("subject_match--->",subject_match)
    subject = subject_match.group(1).strip() if subject_match else ""
    # print("subject--->",subject)
    # Extract Body (after first blank line)
    parts = raw_email.split("\n\n", 1)
    # print("parts--->",parts)
    body = parts[1] if len(parts) > 1 else ""
    # print("body---->  ------->",body)
    # Remove extra whitespace
    body = re.sub(r"\s+", " ", body).strip()
    # print("body---->",body)
    # print()
    # print("//////////////////////////////////////////////")


    return subject, body

In [ ]:
def clean_body(body):
    # Remove quoted replies
    # print("clean_body -........................................")
    # print("body---->  ------->",body)

    body = "\n".join(
        line for line in body.split("\n")
        if not line.strip().startswith(">")
    )
    # print("body---->  ------->",body)

    # Remove forwarded/reply chains
    body = re.split(r"-----Original Message-----", body)[0]
    # print("body---->  ------->",body)

    # Remove signatures (basic)
    body = re.split(r"(Regards|Thanks|Sincerely)", body)[0]
    # print("body---->  ------->",body)

    # Final cleanup
    body = re.sub(r"\s+", " ", body).strip()
    # print("body---->  ------->",body)

    return body

In [ ]:
def to_training_sample(subject, body, metadata=None):
    # email_text = f"{subject}\n\n{body}" if subject else body
    # email_text = f"{subject}\n{metadata}\n\n{body}" if subject and metadata else body
    email_text = "\n".join(filter(None, [subject, metadata, "", body]))

    return email_text

In [ ]:
def process_email(raw_email):
    # subject, body = parse_enron_email(raw_email)
    # body = clean_body(body)
    subject, body,metadata = parse_enron_email_tmp(raw_email)

    prompt = to_training_sample(subject, body)

    return prompt

In [ ]:
n = random.randint(0,517401)
pmt= process_email(enron_np[n][1])
# print(enron_np[n][1])
print("---------------------------------")
print(pmt)

---------------------------------
Policy Narrative Draft
Attached for you review and comments is a revised policy narrative. We have incoprorated the recommendations from the N-Form group and the ideas provided by Neil and his team. Please note the new title: Talent Deployment as well as other changes. As always please feel free to shoot your comments back.


In [ ]:
for i in range (1):
    n = random.randint(0,517401)
    pmt= process_email(enron_np[n][1])
    print(enron_np[n][1])
    print("---------------------------------")
    print(pmt)
    print()
    print(f"=============== {i} =============")

Message-ID: <15955202.1075846287451.JavaMail.evans@thyme>
Date: Mon, 12 Jun 2000 01:51:00 -0700 (PDT)
From: maureen.mcvicker@enron.com
To: w/assts <govt.>
Subject: Memo on Regulatory Risks
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: Maureen McVicker
X-To: Govt Affairs (w/Assts)
X-cc: 
X-bcc: 
X-Folder: \Steven_Kean_Dec2000_1\Notes Folders\Calendar\Untitled
X-Origin: KEAN-S
X-FileName: skean.nsf

Please see the attached memo.












---------------------------------
Memo on Regulatory Risks
Please see the attached memo.

=============== 0 =============


In [ ]:
from openai import OpenAI
import os

# kaggle_api = "KGAT_88615a4948b07c40fbd1086370ec5d75"

# GROQ_API_KEY = "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"
# GROQ_API_KEY = "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"
# GROQ_API_KEY = "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"

In [ ]:

client = OpenAI(
    api_key=GROQ_API_KEY,
    base_url="https://api.groq.com/openai/v1",
)

response = client.responses.create(
    input="hey",
    model="openai/gpt-oss-20b",
)
print(response.output_text)


Hey there! How can I help you today?


In [ ]:
def call_llm(payload):

    response = client.responses.create(
        input=payload,
        model="openai/gpt-oss-20b",
    )

    return response

In [ ]:
def generate_label(email_text):
    prompt = f"""
### Instruction:
    Analyze the email and return JSON with:
    - category (Spam, Meeting, Work, Personal, Promotion)
    - priority (High, Medium, Low)
    - entities (date in YYYY-MM-DD if possible, time in HH:MM 24hr format, people as list of names)
    - summary (short and clear)
    - action (schedule_meeting, reply_with_availability, draft_reply, ignore, notify, none)
    - reply (natural email-style response)

Email:
{email_text}

Return JSON only.

"""
    response = call_llm(prompt)
    return response.output_text

In [ ]:
n = random.randint(0,517401)
print(n)
email= process_email(enron_np[n][1])
# email = enron_np[n][1]
print(enron_np[n][1])
print("---------------------------------")
print(email)
print("---------------------------------")
response = generate_label(email)
print(response)

98242
Message-ID: <7700559.1075855476098.JavaMail.evans@thyme>
Date: Fri, 3 Nov 2000 10:38:00 -0800 (PST)
From: david.forster@enron.com
To: eol.wide@enron.com
Subject: Change to Stack Manager
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Transfer-Encoding: 7bit
X-From: David Forster
X-To: EOL Traders World Wide
X-cc: 
X-bcc: 
X-Folder: \Frank_Ermis_Dec2000\Notes Folders\All documents
X-Origin: Ermis-F
X-FileName: fermis.nsf

Effective immediately, there is a change to the Stack Manager which will 
affect Inactive Children.

Inactive Children with links to Parent products will not have their 
calculated prices updated until the Child product is Activated.

When the Child Product is activated, the price will be recalculated and 
updated BEFORE it is displayed on the web.

This means that if you are inputting a basis price on a Child product, you 
will not see the final, calculated price until you Activate the product, at 
which time the customer will also see it.



In [ ]:
import json
import time

dataset = []

for i in range(1000):
    n = random.randint(0, 517401)

    email = process_email(enron_np[n][1])

    response = generate_label(email)  # your LLM output

    sample = {
        "instruction": "Analyze the email and return structured JSON",
        "input": email,
        "output": response
    }

    dataset.append(sample)


RateLimitError: Error code: 429 - {'error': {'message': 'Rate limit reached for model `openai/gpt-oss-20b` in organization `org_01kpfe9cspfqd81qebea6y2jwb` service tier `on_demand` on tokens per day (TPD): Limit 200000, Used 198551, Requested 1550. Please try again in 43.632s. Need more tokens? Upgrade to Dev Tier today at https://console.groq.com/settings/billing', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}

In [ ]:
print(len(dataset))


18


In [ ]:
with open("/kaggle/working/enron_structured1.jsonl", "a") as f:
    for item in dataset:
        f.write(json.dumps(item) + "\n")

In [ ]:
df = pd.read_json("/kaggle/working/enron_structured1.jsonl", lines=True)
data = np.array(df)


In [ ]:
print(len(data))

67


In [ ]:
print(dataset[11])

{'instruction': 'Analyze the email and return structured JSON', 'input': 'RE: 814 enrollment transaction training\nThe class is now closed. We have 50 attendees. Thank you Karen Bergman Senior Client Services Representative 512-248-3171 kbergman@ercot.com', 'output': '```json\n{\n  "category": "Work",\n  "priority": "Medium",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": ["Karen Bergman"]\n  },\n  "summary": "Training class now closed with 50 attendees.",\n  "action": "ignore",\n  "reply": ""\n}\n```'}


In [ ]:
df = pd.read_json("/kaggle/input/datasets/divyamgupta04/enron-struct-data/enron_structured (4).jsonl", lines=True)
data = np.array(df)
data[3]

array(['Analyze the email and return structured JSON',
       'EOL Issues\nWillie:',
       '{\n  "category": "Work",\n  "priority": "Medium",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": ["Willie"]\n  },\n  "summary": "Email regarding End‑of‑Life (EOL) issues",\n  "action": "reply",\n  "reply": "Hi,\\n\\nThanks for bringing up the EOL issues. Could you please provide a bit more detail or the specific components you’re referring to? I’ll look into it and get back to you shortly.\\n\\nBest regards,\\n[Your Name]"\n}\n'],
      dtype=object)

In [ ]:
import json
# data_json = json.loads(dataset[0])
out = data[3][2]
out_json = json.loads(out)

In [ ]:
out_json

{'category': 'Work',
 'priority': 'Medium',
 'entities': {'date': None, 'time': None, 'people': ['Willie']},
 'summary': 'Email regarding End‑of‑Life (EOL) issues',
 'action': 'reply',
 'reply': 'Hi,\n\nThanks for bringing up the EOL issues. Could you please provide a bit more detail or the specific components you’re referring to? I’ll look into it and get back to you shortly.\n\nBest regards,\n[Your Name]'}

In [ ]:
def prompt(email_text):
    return f"""
Analyze the email and return JSON with:
- category
- priority
- entities (date, time, people)
- summary
- action
- reply

categories = [
    "Spam",
    "Meeting",
    "Work",
    "Personal",
    "Promotion"
]

Email:
{email_text}

Return JSON only.

------------------------------
A sample example of an email
------------------------------

Email:
Hey, can we meet tomorrow at 3 PM?

Response:
{{
  "category": "Meeting",
  "priority": "High",
  "entities": {{
    "date": "tomorrow",
    "time": "15:00",
    "people": [
    {{
      "name": "null",
      "email": null
    }}
  ]
  }},
  "summary": "Meeting request for tomorrow at 3 PM",
  "action": "schedule_meeting",
  "reply": "Sure, 3 PM works for me."
}}
"""


from google import genai

key="AIzaSyDJK3KzG1MxgmqfabVPc3ojHK5DT4vHHY0"

client = genai.Client(api_key="AIzaSyDJK3KzG1MxgmqfabVPc3ojHK5DT4vHHY0")

email= process_email(enron_np[n][1])
pmt = prompt(email)
response = client.models.generate_content(
    model="gemini-3-flash-preview",
    contents=pmt,
)

print(response.text)

{
  "category": "Personal",
  "priority": "Medium",
  "entities": {
    "date": "Tue, 7 Mar 2000",
    "time": "05:50:00",
    "people": [
      {
        "name": "Tana Jones",
        "email": "tana.jones@enron.com"
      },
      {
        "name": "null",
        "email": "lofeco@eul.net"
      }
    ]
  },
  "summary": "Tana Jones is requesting a bid for a list of seven maintenance and repair tasks for a downstairs apartment, including painting cabinets, fixing drawers, and replacing hardware.",
  "action": "provide_bid",
  "reply": "Thank you for the detailed list, Tana. I will review these items and provide you with a bid for the work as soon as possible."
}


#  **DATA**

**Dataset Cleaning and final fomating in**

{
  "messages": [
    {
      "role": "system",
      "content": "You are an email assistant. Extract structured information and return ONLY valid JSON."
    },
    {
      "role": "user",
      "content": "Analyze the email:\n\nRE: 814 enrollment transaction training\nThe class is now closed..."
    },
    {
      "role": "assistant",
      "content": "{\n  \"category\": \"Work\",\n  \"priority\": \"Medium\",\n  ...\n}"
    }
  ]
}

In [ ]:
import json
import re

def clean_json_output(text):
    # Remove ```json ``` wrappers if present
    text = re.sub(r"```json\s*", "", text)
    text = re.sub(r"```", "", text)

    # Strip whitespace
    text = text.strip()

    # Try to extract JSON block if extra text exists
    try:
        # Find first { ... } block
        match = re.search(r"\{.*\}", text, re.DOTALL)
        if match:
            text = match.group(0)

        return json.loads(text)
    except Exception as e:
        return None


def convert_row(row):
    system_prompt = "You are an email assistant. Extract structured information and return ONLY valid JSON."

    user_prompt = f"Analyze the email:\n\n{row[1]}"

    cleaned = clean_json_output(row[2])

    if cleaned is None:
        return None  # skip bad rows

    return {
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
            {"role": "assistant", "content": json.dumps(cleaned, indent=2)}
        ]
    }

In [ ]:
df = pd.read_json("/kaggle/input/datasets/divyamgupta04/enron-struct-data/enron_structured (4).jsonl", lines=True)
output_semi_enron = np.array(df)

In [ ]:
output_file = "/kaggle/working/enron_chat.jsonl"

with open(output_file, "a") as f:
    for row in output_semi_enron:   # assuming each row is like your array
        # print(row)
        converted = convert_row(row)
        # print(converted)
        # break
        if converted:
            f.write(json.dumps(converted) + "\n")

In [ ]:
df = pd.read_json("/kaggle/working/enron_chat.jsonl", lines=True)
data = np.array(df)

In [ ]:
print(len(data))

829


In [ ]:
data[0]

array([list([{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': 'Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch'}, {'role': 'assistant', 'content': '{\n  "category": "Meeting",\n  "priority": "Low",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": [\n      "Fletch"\n    ]\n  },\n  "summary": "Agreement to proceed with the proposed plan.",\n  "action": "reply_with_availability",\n  "reply": "Hi,\\n\\nSounds great to me. Could you let me know the next steps or the meeting details?\\n\\nThanks,\\nFletch"\n}'}])],
      dtype=object)

In [ ]:


def extract_email_content(user_text):
    # Extract only the email part after the prompt
    parts = user_text.split("Analyze the email:\n\n", 1)
    return parts[1] if len(parts) > 1 else user_text


def normalize_text(text):
    text = text.lower()
    text = re.sub(r"re:|fw:", "", text)   # optional cleanup
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def remove_duplicates(input_file, output_file):
    seen = set()

    with open(input_file, "r") as fin, open(output_file, "w") as fout:
        for line in fin:
            try:
                data = json.loads(line)

                # get user message
                user_msg = next(
                    msg["content"] for msg in data["messages"] if msg["role"] == "user"
                )

                # extract actual email
                email_text = extract_email_content(user_msg)

                key = normalize_text(email_text)

                if key in seen:
                    continue  # ❌ drop duplicate

                seen.add(key)
                fout.write(json.dumps(data) + "\n")

            except Exception:
                continue

In [ ]:
remove_duplicates(
    "enron_chat.jsonl",
    "enron_chat_clean.jsonl"
)

In [ ]:
df = pd.read_json("/kaggle/working/enron_chat_clean.jsonl", lines=True)
data = np.array(df)

In [ ]:
print(len(data))

827


In [ ]:
print(data[0])

[list([{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': 'Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch'}, {'role': 'assistant', 'content': '{\n  "category": "Meeting",\n  "priority": "Low",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": [\n      "Fletch"\n    ]\n  },\n  "summary": "Agreement to proceed with the proposed plan.",\n  "action": "reply_with_availability",\n  "reply": "Hi,\\n\\nSounds great to me. Could you let me know the next steps or the meeting details?\\n\\nThanks,\\nFletch"\n}'}])]


In [ ]:
data[1][0]

[{'role': 'system',
  'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'},
 {'role': 'user',
  'content': "Analyze the email:\n\nConference Call\nBryan, I am trying to set up a conference call 30 minutes with Andy Kalotay. I worked with Andy at Salomon; he is now running his consulting firm in NYC and would like to talk to us about marketing a credit product he helped to develop. The info is attached below. Shirley will call you in the morning. Vince ******************************************************************************* S&P Credit Indices S&P Credit Indices reflect daily movements in credit spread levels within two sectors of the U.S. bond market. S&P Credit Indices are quoted in basis points over the U.S. Treasury yield curve once daily. Credit spread movements within the investment grade spectrum are represented in the S&P U.S. Industrial Investment Grade Credit Index while movements within the speculative grade rating spectru

In [ ]:
data[0][0][0]['content']

'You are an email assistant. Extract structured information and return ONLY valid JSON.'

In [ ]:
train = {
        "messages": [
            {"role": "system", "content": data[0][0][0]['content']},
            {"role": "user", "content": data[0][0][1]['content']},
            {"role": "assistant", "content": data[0][0][2]['content']}
        ]
    }

In [ ]:
print(train)

{'messages': [{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': 'Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch'}, {'role': 'assistant', 'content': '{\n  "category": "Meeting",\n  "priority": "Low",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": [\n      "Fletch"\n    ]\n  },\n  "summary": "Agreement to proceed with the proposed plan.",\n  "action": "reply_with_availability",\n  "reply": "Hi,\\n\\nSounds great to me. Could you let me know the next steps or the meeting details?\\n\\nThanks,\\nFletch"\n}'}]}


In [ ]:
import json

def fix_structure(input_file, output_file):
    with open(input_file, "r") as fin, open(output_file, "w") as fout:
        for line in fin:
            try:
                data = json.loads(line)

                # Case 1: [[{...}]]
                if isinstance(data, list) and len(data) > 0:
                    inner = data[0]

                    # If inner is already list of messages
                    if isinstance(inner, list):
                        fixed = {"messages": inner}

                    else:
                        continue
                else:
                    continue

                fout.write(json.dumps(fixed) + "\n")

            except Exception:
                continue

In [ ]:
with open("train_enron_chat_final.jsonl", "a") as f:
    for i in range (len(data)):
        train = {
            "messages": [
                {"role": "system", "content": data[i][0][0]['content']},
                {"role": "user", "content": data[i][0][1]['content']},
                {"role": "assistant", "content": data[i][0][2]['content']}
            ]
        }
        f.write(json.dumps(train) + "\n")

In [ ]:
with open("/kaggle/working/enron_chat_clean.jsonl", "r") as f:
    cnt=0
    for line in f:
        training_data = json.loads(line)
        print(training_data)
        print()
        if cnt>5:
            break
        cnt = cnt+1

{'messages': [{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': 'Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch'}, {'role': 'assistant', 'content': '{\n  "category": "Meeting",\n  "priority": "Low",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": [\n      "Fletch"\n    ]\n  },\n  "summary": "Agreement to proceed with the proposed plan.",\n  "action": "reply_with_availability",\n  "reply": "Hi,\\n\\nSounds great to me. Could you let me know the next steps or the meeting details?\\n\\nThanks,\\nFletch"\n}'}]}

{'messages': [{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': "Analyze the email:\n\nConference Call\nBryan, I am trying to set up a conference call 30 minutes with Andy Kalotay. I worked with Andy at Salomon; he is now running h

{'messages': [{'role': 'system', 'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'}, {'role': 'user', 'content': 'Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch'}, {'role': 'assistant', 'content': '{\n  "category": "Meeting",\n  "priority": "Low",\n  "entities": {\n    "date": null,\n    "time": null,\n    "people": [\n      "Fletch"\n    ]\n  },\n  "summary": "Agreement to proceed with the proposed plan.",\n  "action": "reply_with_availability",\n  "reply": "Hi,\\n\\nSounds great to me. Could you let me know the next steps or the meeting details?\\n\\nThanks,\\nFletch"\n}'}]}


# **Training**

In [ ]:
map={}
with open("/kaggle/input/train-enron/train_enron_chat_final.jsonl", "r") as f:
    cnt=0
    for line in f:
        training_data = json.loads(line)
        # print(training_data['messages'][2]['content'])
        tmp = json.loads(training_data['messages'][2]['content'])
        # print(tmp['action'])
        map[tmp['action']] = map.get(tmp['action'], 0) + 1
        print()
        # if cnt>3:
        #     break
        cnt = cnt+1

In [ ]:
print(map)

{'reply_with_availability': 131, 'draft_reply': 363, 'reply': 15, 'none': 172, 'ignore': 87, 'schedule_meeting': 40, 'notify': 20}


In [ ]:
!pip install -U transformers accelerate bitsandbytes peft trl datasets

In [ ]:
import torch
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    BitsAndBytesConfig,
    TrainingArguments
)
from peft import LoraConfig, get_peft_model
from datasets import Dataset
from trl import SFTTrainer

In [ ]:
# -----------------------
# Model
# -----------------------
model_id = "microsoft/Phi-3.5-mini-instruct"

tokenizer = AutoTokenizer.from_pretrained(model_id)

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4"
)

# model = AutoModelForCausalLM.from_pretrained(
#     model_id,
#     quantization_config=bnb_config,
#     device_map="auto",
#     offload_folder="offload",
#     offload_state_dict=True
# )
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    device_map="auto",
    torch_dtype=torch.float16,   # ✅ force FP16
    offload_folder="offload",
    offload_state_dict=True
)

# IMPORTANT memory fixes
model.config.use_cache = False
model.gradient_checkpointing_enable()


`torch_dtype` is deprecated! Use `dtype` instead!


Loading weights:   0%|          | 0/195 [00:00<?, ?it/s]

In [ ]:
# -----------------------
# LoRA (PEFT)
# -----------------------
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["qkv_proj", "o_proj"],  # ✅ correct for Phi
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)

model = get_peft_model(model, lora_config)


In [ ]:
model = model.to(torch.float16)

In [ ]:
for name, module in model.named_modules():
    print(name)


base_model
base_model.model
base_model.model.model
base_model.model.model.embed_tokens
base_model.model.model.layers
base_model.model.model.layers.0
base_model.model.model.layers.0.self_attn
base_model.model.model.layers.0.self_attn.o_proj
base_model.model.model.layers.0.self_attn.o_proj.base_layer
base_model.model.model.layers.0.self_attn.o_proj.lora_dropout
base_model.model.model.layers.0.self_attn.o_proj.lora_dropout.default
base_model.model.model.layers.0.self_attn.o_proj.lora_A
base_model.model.model.layers.0.self_attn.o_proj.lora_A.default
base_model.model.model.layers.0.self_attn.o_proj.lora_B
base_model.model.model.layers.0.self_attn.o_proj.lora_B.default
base_model.model.model.layers.0.self_attn.o_proj.lora_embedding_A
base_model.model.model.layers.0.self_attn.o_proj.lora_embedding_B
base_model.model.model.layers.0.self_attn.o_proj.lora_magnitude_vector
base_model.model.model.layers.0.self_attn.qkv_proj
base_model.model.model.layers.0.self_attn.qkv_proj.base_layer
base_model.

In [ ]:
from datasets import Dataset
data = Dataset.from_json("/kaggle/input/train-enron/train_enron_chat_final.jsonl")

In [ ]:
len(data)

828

In [ ]:
# data = [...]  # your 750 samples

def format_example(example):
    text = ""
    for msg in example["messages"]:
        text += f"{msg['role']}: {msg['content']}\n"
    return {"text": text}

dataset = Dataset.from_list(data)
dataset = dataset.map(format_example)

Map:   0%|          | 0/828 [00:00<?, ? examples/s]

In [ ]:
def tokenize(example):
    return tokenizer(
        example["text"],
        truncation=True,
        max_length=256,   # keep this low for Kaggle
        padding="max_length"
    )

dataset = dataset.map(tokenize, batched=True)

Map:   0%|          | 0/828 [00:00<?, ? examples/s]

In [ ]:
from transformers import TrainingArguments

training_args = TrainingArguments(
    output_dir="./phi-email-agent",
    per_device_train_batch_size=1,
    gradient_accumulation_steps=8,
    learning_rate=2e-4,
    num_train_epochs=3,
    logging_steps=10,
    save_strategy="epoch",
    fp16=False,
    bf16=False,
    report_to="none",
    optim="adamw_torch"   # ✅ IMPORTANT (avoids fused/AMP paths)
)

In [ ]:
from trl import SFTTrainer
from transformers import DataCollatorForLanguageModeling


data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=False
)

trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    data_collator=data_collator,
    args=training_args
)

trainer.train()

Step,Training Loss
10,2.667536
20,1.750552
30,1.491472
40,1.381253
50,1.430868
60,1.276501
70,1.297956
80,1.336189
90,1.315859
100,1.260496


TrainOutput(global_step=312, training_loss=1.2937410779488392, metrics={'train_runtime': 1446.4606, 'train_samples_per_second': 1.717, 'train_steps_per_second': 0.216, 'total_flos': 1.4239223903158272e+16, 'train_loss': 1.2937410779488392})

In [ ]:
model.save_pretrained("phi-email-agent-lora")
tokenizer.save_pretrained("phi-email-agent-lora")

('phi-email-agent-lora/tokenizer_config.json',
 'phi-email-agent-lora/chat_template.jinja',
 'phi-email-agent-lora/tokenizer.json')

In [ ]:
import shutil

shutil.make_archive("/content/phi-email-agent-lora", 'zip', "/content/phi-email-agent-lora")

'/content/phi-email-agent-lora.zip'

# **Inference**

In [ ]:
import zipfile

zip_path = "/content/phi-email-agent-lora.zip"
extract_path = "/content/phi-email-agent-lora"

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_path)

In [ ]:
!pip install -U bitsandbytes accelerate transformers peft

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 13.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 130.6 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 5.0.0
    Uninstalling transformers-5.0.0:
      Successfully uninstalled transformers-5.0.0


In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from peft import PeftModel

model_id = "microsoft/Phi-3.5-mini-instruct"

tokenizer = AutoTokenizer.from_pretrained(model_id)

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4"
)

base_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    device_map="auto"
)

model = PeftModel.from_pretrained(
    base_model,
    "/content/phi-email-agent-lora"
)

model.eval()

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json: 0.00B [00:00, ?B/s]

[transformers] This model config has set a `rope_parameters['original_max_position_embeddings']` field, to be used together with `max_position_embeddings` to determine a scaling factor. Please set the `factor` field of `rope_parameters`with this ratio instead -- we recommend the use of this field over `original_max_position_embeddings`, as it is compatible with most model architectures.


tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/306 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/195 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/195 [00:00<?, ?B/s]

PeftModelForCausalLM(
  (base_model): LoraModel(
    (model): Phi3ForCausalLM(
      (model): Phi3Model(
        (embed_tokens): Embedding(32064, 3072, padding_idx=32000)
        (layers): ModuleList(
          (0-31): 32 x Phi3DecoderLayer(
            (self_attn): Phi3Attention(
              (o_proj): lora.Linear4bit(
                (base_layer): Linear4bit(in_features=3072, out_features=3072, bias=False)
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.05, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=3072, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=3072, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
                (lora_magnitude_vector): ModuleDict()
              )
             

In [ ]:
prompt = tokenizer.apply_chat_template(
    [
        {
            'role': 'system',
            'content': 'You are an email assistant. Extract structured information and return ONLY valid JSON.'
        },
        {
            'role': 'user',
            'content': "Analyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch"
        }
    ],
    tokenize=False,
    add_generation_prompt=True
)

In [ ]:
prompt

'<|system|>\nYou are an email assistant. Extract structured information and return ONLY valid JSON.<|end|>\n<|user|>\nAnalyze the email:\n\nRE: Cinergy vs COMED phy spread\nsounds great to me. fletch<|end|>\n<|assistant|>\n'

In [ ]:
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

with torch.no_grad():
    outputs = model.generate(
        **inputs,
        max_new_tokens=1000,
        temperature=0.2,   # low → more deterministic JSON
        do_sample=False
    )

result = tokenizer.decode(outputs[0], skip_special_tokens=True)
print(result)

[transformers] The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.


You are an email assistant. Extract structured information and return ONLY valid JSON. Analyze the email:

RE: Cinergy vs COMED phy spread
sounds great to me. fletch {
  "category": "Work",
  "priority": "Medium",
  "entities": {
    "date": null,
    "time": null,
    "people": [
      "Fletch"
    ]
  },
  "summary": "Sender agrees with the phy spread strategy and expresses approval.",
  "action": "draft_reply",
  "reply": "Hi Fletch,\n\nThanks for your quick response. I\u2019m glad to hear that you\u2019re on board with the phy spread strategy. Let\u2019s schedule a brief call to discuss the next steps and any potential concerns.\n\nBest regards,\n[Your Name]"
}



In [ ]:
response = result.split("assistant:")[-1].strip()
print(response)

You are an email assistant. Extract structured information and return ONLY valid JSON. Analyze the email:

RE: Cinergy vs COMED phy spread
sounds great to me. fletch {
  "category": "Work",
  "priority": "Medium",
  "entities": {
    "date": null,
    "time": null,
    "people": [
      "Fletch"
    ]
  },
  "summary": "Sender agrees with the phy spread strategy and expresses approval.",
  "action": "draft_reply",
  "reply": "Hi Fletch,\n\nThanks for your quick response. I\u2019m glad to hear that you\u2019re on board with the phy spread strategy. Let\u2019s schedule a brief call to discuss the next steps and any potential concerns.\n\nBest regards,\n[Your Name]"
}


In [ ]:
import re

def extract_json(text):
    match = re.search(r"\{.*\}", text, re.DOTALL)
    return match.group(0) if match else None

json_str = extract_json(result)
print(json_str)

{
  "category": "Work",
  "priority": "Medium",
  "entities": {
    "date": null,
    "time": null,
    "people": [
      "Fletch"
    ]
  },
  "summary": "Sender agrees with the phy spread strategy and expresses approval.",
  "action": "draft_reply",
  "reply": "Hi Fletch,\n\nThanks for your quick response. I\u2019m glad to hear that you\u2019re on board with the phy spread strategy. Let\u2019s schedule a brief call to discuss the next steps and any potential concerns.\n\nBest regards,\n[Your Name]"
}


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive
